### ISSUE TITLE
Wallpanel crashes with current Chrome desktop user agent string

### ISSUE THREAD

**Issue by habitoti** [thanksmister/wallpanel-android#287]
**Describe the bug**
Wanted to display content with desktop browser rendering (not mobile). So I entered the latest Chrome standard user agent:
```
Mozilla/5.0 (Linux; Android 10) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Mobile Safari/537.36
```
The WebView did no longer do anything. So I stopped and restarted Wallpanel, but it now crashes straight on startup and doesn't even let me get into settings to revert the agent.

**To Reproduce**

**Smartphone (please complete the following information):**
 - Lenovo M8 tablet

**Complete explanation ;-)
Amazon Prime day brought me a half priced Lenovo M8 tablet, which would make a great additional wall panel (used Nexus 7 so far). Other than for the Nexus, the System WebView asks for mobile content per default, which is a bit too large. The experimental Wallpanel zoom feature does nothing for me, unfortunately, but using Chrome to display the website I noticed that when requesting the desktop version, it fits perfectly. To achieve the same with Wallpanel, I wanted to set the current user agent so that it behaves like a desktop browser.

The **generally preferred option** would be, of course, if Wallpanel would introduce a switch to programmatically force the WebView to optionally request a desktop version (see also [here](https://stackoverflow.com/questions/14688030/setting-webview-to-view-desktop-site-and-not-mobile-site)...).

**Comment by habitoti** [comment]
Also: if I open settings via the popup on the home screen icon, I cannot delete the user agent. It doesn't accept to clear the complete input (and continues to use the existing)

**Comment by habitoti** [comment]
The journey continues ;-)
I deleted Wallpanel and reinstalled it. Now the above string no longer crashes Wallpanel, however it has no effect obviously. While Chrome renders the page correctly when requested as desktop version, Wallpanel just uses mobile rendering. See below examples (note: while chrome still obstructs parts with header & footer, the desktop version has the correct size below it)

Chrome/Desktop:

![1411823D-A268-4B22-9634-33BB893F8220_1_105_c](https://user-images.githubusercontent.com/220421/122944074-cedb9c80-d377-11eb-920c-f7fc42d0b7e2.jpeg)

Chrome/Mobile:

![F97E018D-4716-40FF-ADE0-8FFEA39B9382_1_105_c](https://user-images.githubusercontent.com/220421/122944126-db5ff500-d377-11eb-9e41-825dca389019.jpeg)

Wallpanel:

![704500BA-60FC-41C9-9019-A66EE032A83A_1_105_c](https://user-images.githubusercontent.com/220421/122944222-eb77d480-d377-11eb-8485-2a65a2e05937.jpeg)

**Comment by thanksmister** [comment]
The stack post you linked to has a different user agent than the one you provided in your ticket.   

`
Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.9.0.4) Gecko/20100101 Firefox/4.0
`

The application already has the ability to set a user agent.   I don't think the app needs to be changed to accomodate this request, this seems more about how WebView handles user agents.  I assume it might not work and depends also on your version of WebView.    

To be clear, WallPanel does not use Chrome, it will not be the same as Chrome browser, it will always be WebView, which has it's own page rendering ability.   You cannot compare how Chrome renders a page as a desktop and how WebView does.  Also without the correct user agent or settings, maybe WebView will just ignore your attempts to render a desktop version of the site.

Also, this seems like a problem you can address in the actual web page, this is why its rendering the mobile version, the webpage is built to do so.    I mean, just a thought, but update or override this on the webpage.

**Comment by habitoti** [comment]
Well, I know. I just quoted the stack overflow page, as beside setting the user agent it seems to require some more programmatic settings, like
```
webSettings.setUseWideViewPort(true);
webSettings.setLoadWithOverviewMode(true);
webSettings.setSupportZoom(true);
webSettings.setBuiltInZoomControls(true);
```
...which is -- in addition to setting the UA -- probably what Chrome does once you switch it to "Desktop" mode.

The server site (which is Homeassistant) unfortunately has no way to influence how dashboards are provided. 

Anyway, I understand your point. I just don't know how to make WebView display the page in correct format (as also the Zoom seemingly doesn't work).

Can you anyway please check the deletion of the user agent. That really doesn't seem to work. Setting it back to an empty string will fail and be ignored.

**Comment by thanksmister** [comment]
Have you already looked at the code to see if these options were already set on the WebView?   

```
webSettings.setUseWideViewPort(true);
webSettings.setLoadWithOverviewMode(true);
webSettings.setSupportZoom(true);
webSettings.setBuiltInZoomControls(true);
```

I am pretty sure there is a way to hack HA to do this, you might want to post on the HA community site to see if anyone has already made a hack for this or there is a way to hack this yourself.

**Comment by habitoti** [comment]
That are methods of the WebView. Don't think you can call them through JS from loaded content.

**Comment by thanksmister** [comment]
These are our current settings: 

```
 if(webSettings == null) {
            webSettings = mWebView.settings
        }
        webSettings?.javaScriptEnabled = true
        webSettings?.domStorageEnabled = true
        webSettings?.databaseEnabled = true
        webSettings?.saveFormData = true
        webSettings?.javaScriptCanOpenWindowsAutomatically = true
        webSettings?.setAppCacheEnabled(true)
        webSettings?.allowFileAccess = true
        webSettings?.allowFileAccessFromFileURLs = true
        webSettings?.allowContentAccess = true
        webSettings?.setSupportZoom(true)
        webSettings?.loadWithOverviewMode = true
        webSettings?.useWideViewPort = true
        webSettings?.pluginState = WebSettings.PluginState.ON
        webSettings?.setRenderPriority(WebSettings.RenderPriority.HIGH);
       // webSettings?.cacheMode = WebSettings.LOAD_NO_CACHE;

        if (userAgent.isNotEmpty()) {
            webSettings?.userAgentString = userAgent
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webSettings?.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }
```

**Comment by thanksmister** [comment]
> That are methods of the WebView. Don't think you can call them through JS from loaded content.

You can probably disable HA from delivering a web version of the page, that is what I mean.  This is part of the web page css, to offer a mobile version of the page.  If this could be overwritten, then the page would always render as a desktop version.

**Comment by habitoti** [comment]
Yeah, need to check with the HA folks then. Still confused why WebView renders exactly like Chrome/Mobile on the Lenovo tab, even while your settings from above look good for desktop, and using an appropriate UA. Plus: works also right on the Nexus w/o any mod. of UA, using same WebView. No clue how this fits together...

**Comment by thanksmister** [comment]
You are only missing `webSettings.setBuiltInZoomControls(true);` and from what I can find, I don't think this is a requirement to make the user agent work.   I think HA is just ignoring your user agent or maybe we are not applying it in the application.  I will look at what's going on with setting/resetting the user agent.

**Comment by habitoti** [comment]
No, don't think either

**Comment by habitoti** [comment]
I think the agent is set correctly, because setting a "bad" one will raise an error ("unsupported browser version"...). Unfortunately, though, I can't reset it then to an empty one ;-)

**Comment by thanksmister** [comment]
The code prevents the user agent from being empty, so once you set it, this is why you cannot just set it to empty.  I can fix this, but what you can do now for your testing is just reset the user agent to something normal for WebView.  

```
 PREF_SETTINGS_USER_AGENT -> {
                val value = userAgentPreference.text.orEmpty()
                if (value.isNotEmpty()) {
                    configuration.browserUserAgent = value
                    userAgentPreference.summary = value
                }
            }

```

**Comment by habitoti** [comment]
Regarding HA ignoring it: at least switching Chrome to "Desktop" makes it behave correctly. Not sure if there is more a browser can do than providing the correct UA. At least that is how we use Chrome to toggle between different clients and simulate phones, tablets etc.

**Comment by thanksmister** [comment]
It might  be a matter of finding the correct UA then.  Maybe you can detect and copy the UA that Chrome is setting when requesting "desktop" version?

**Comment by habitoti** [comment]
Yeah, will setup a proxy and check it out

**Comment by thanksmister** [comment]
Great, let me know, would be good if we can get this working and document it.

**Comment by habitoti** [comment]
Ok, so here are the fiddler sniffs:

**Chrome/Mobile:**

Mozilla/5.0 (Linux; Android 10; Lenovo TB-8505F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Mobile Safari/537.36

**Chrome/Desktop:**

Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36

**Webview:**

Mozilla/5.0 (Linux; Android 10; Lenovo TB-8505F Build/QP1A.190711.020; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/91.0.4472.101 Mobile Safari/537.36


But setting the Chrome/Desktop UA doesn't change the rendering in Wallpanel, it's still rendered like Chrome/Mobile :-(
Will have to check with the HA folks whether this can be forced server side...

**Comment by thanksmister** [comment]
I imagine that you can override the HA CSS in some way to prevent it from delivering a mobile version.

**Comment by habitoti** [comment]
Will check...

One more thing for you, though: once I edit the UA to sth. else (because of deletions failure), Wallpanel crashes once you leave settings. Can't be started at all any longer then. I have to delete all app settings and start over to configure everything.

**Comment by habitoti** [comment]
Is there anything else required to make the Zoom feature work? It does nothing for me...

**Comment by thanksmister** [comment]
No, zoom has always been experimental and we already have all the settings for it to work with WebView.  If it doesn't work, it just won't work on your setup or maybe the webpage, but there is nothing else to do.
